### ISSUE TITLE
IPVS not updated in the event of a node failure

### ISSUE THREAD

**Issue by vfrans** [kubernetes/kubernetes#105505]
#### What happened:
When a single node goes down IPs associated with the pods that are on the impacted are not removed from the IPVS service on each remaining nodes leading to traffic being routed to pod's ip that no longer exists.

That being said, IPs are well removed from the service when quering Kubernetes's API (no invalid endpoints are present)

Nodes:
```
kubectl get nodes -o wide
NAME        STATUS     ROLES    AGE   VERSION    INTERNAL-IP     EXTERNAL-IP   OS-IMAGE                KERNEL-VERSION                CONTAINER-RUNTIME
cluster-a   Ready      master   48d   v1.18.10   10.194.10.101   <none>        CentOS Linux 7 (Core)   3.10.0-1160.24.1.el7.x86_64   docker://19.3.12
cluster-b   NotReady   master   48d   v1.18.10   10.194.10.102   <none>        CentOS Linux 7 (Core)   3.10.0-1160.24.1.el7.x86_64   docker://19.3.12
cluster-c   Ready      master   48d   v1.18.10   10.194.10.103   <none>        CentOS Linux 7 (Core)   3.10.0-1160.24.1.el7.x86_64   docker://19.3.12
```

Kube-proxy logs:
```
E1004 08:33:04.015653       1 reflector.go:178] k8s.io/client-go/informers/factory.go:135: Failed to list *v1.Service: Get https://10.194.10.102:6443/api/v1/services?labelSelector=%21service.kubernetes.io%2Fheadless%2C%21service.kubernetes.io%2Fservice-proxy-name&resourceVersion=15545171: dial tcp 10.194.10.102:6443: connect: no route to host
E1004 08:33:15.037905       1 reflector.go:178] k8s.io/client-go/informers/factory.go:135: Failed to list *v1.Endpoints: Get https://10.194.10.102:6443/api/v1/endpoints?labelSelector=%21service.kubernetes.io%2Fheadless%2C%21service.kubernetes.io%2Fservice-proxy-name&resourceVersion=16581945: dial tcp 10.194.10.102:6443: connect: no route to host
```

When node is killed, ips are updated in service:
```
kubectl describe service evs-phoenix-licensing
Name:              evs-phoenix-licensing
Namespace:         ipd-via
Labels:            <none>
Annotations:       <none>
Selector:          evsapp=evs-phoenix-licensing
Type:              ClusterIP
IP:                10.233.62.124
Port:              http  8080/TCP
TargetPort:        8080/TCP
Endpoints:         10.233.64.143:8080,10.233.65.81:8080
Session Affinity:  None
Events:            <none>
```

but not in ipvs:
```
sudo ipvsadm -ln | grep -A 4 10.233.62.124
TCP  10.233.62.124:8080 rr
  -> 10.233.64.143:8080           Masq    1      2          4
  -> 10.233.65.247:8080           Masq    1      0          2
  -> 10.233.66.91:8080            Masq    1      0          2
```

Notice the weight is still 1 on the failed pod's ip: 10.233.66.91:8080 

#### What you expected to happen:
IPVS should be updated on the remaining nodes, removing the impacted pod's ip to ensure no traffic will be routed to pods that were residing on failed nodes.

#### How to reproduce it (as minimally and precisely as possible):
3 nodes cluster, all master+worker nodes running 1.18.10 (using containerd version 1.2.13). Kill a single node (we are using VMware VMs, so just force power off a VM)

#### Environment:
- Kubernetes version (use `kubectl version`):
Kubernetes 1.18.10 (installed using kubespray)
containerd.io-1.2.13-3.2.el7.x86_64
Linux 3.10.0-1160.24.1.el7.x86_64
Flannel as a network plugin

- Cloud provider or hardware configuration:
VMware Cluster on premises
- OS (e.g: `cat /etc/os-release`): CentOS Linux release 7.9.2009 (Core)
- Kernel (e.g. `uname -a`): Linux cluster-a 3.10.0-1160.24.1.el7.x86_64 x86_64 x86_64 GNU/Linux
- Install tools: Kubespray
- Network plugin and version (if this is a network-related bug): flannel 
- Others:

**Comment by cyclinder** [comment]
In the same situation, is it normal if it is in iptables mode?

**Comment by vfrans** [comment]
Never tried iptables mode TBH, didn't even know I could switch ± easily.

**Comment by cyclinder** [comment]
What you are talking about is that when the node goes down, the ipvs service is updated normally(removing  endpoints  that were residing on failed nodes), but the ipvs rules are not updated. right?

**Comment by cyclinder** [comment]
What you are talking about is that when the node goes down, the ipvs service is updated normally(removing  endpoints  that were residing on failed nodes), but the ipvs rules are not updated. right?

**Comment by vfrans** [comment]
No I meant that, when the node goes down, the Kubernetes service (kubectl describe svc) is updated normally:
```
kubectl describe service phoenix-licensing
Name:             phoenix-licensing
Namespace:         myns
Labels:            <none>
Annotations:       <none>
Selector:         app=phoenix-licensing
Type:              ClusterIP
IP:                10.233.62.124
Port:              http  8080/TCP
TargetPort:        8080/TCP
Endpoints:         10.233.64.143:8080,10.233.65.81:8080
Session Affinity:  None
Events:            <none>
```

But the IPVS service (I don't know what the IPVS rules are) is not:
```
sudo ipvsadm -ln | grep -A 4 10.233.62.124
TCP  10.233.62.124:8080 rr
  -> 10.233.64.143:8080           Masq    1      2          4
  -> 10.233.65.247:8080           Masq    1      0          2
  -> 10.233.66.91:8080            Masq    1      0          2
```

**Comment by cyclinder** [comment]
yeah,what you said is consistent with what I think,ipvs service is ipvs rule,but ipvs rule is not updated

**Comment by cyclinder** [comment]
I will try to reproduce this problem in my environment，and find out the cause of this problem

**Comment by khenidak** [comment]
The behavior is the same across all proxies.

The proxies on healthy node does not look at other unhealthy nodes. it looks at objects from api-server. Until the node is marked unhealthy and pods are replaced/deleted they will be still reflected in proxies on other nodes. To help with that situation:
1. Use liveness and readiness probes.
2. short-circuit the node lifecycle controller (use less interval etc).

**Comment by aojea** [comment]
> It seems logical to me that the api servers on the surviving nodes try and fail to reach the api server on the powered off node isn’t it?

why they should try to reach an api server in a different node?

Each kubeproxy has its own kubeconfig, where the api server address is configured, if kube-proxy is not able to get to the apiserver, he is blind and don't know what happens.

The test is node-b goes down, kube-proxy in node-a (using an apiserver that is not node-b) should have the correct information

**Comment by vfrans** [comment]
I'm unsure but I think that Kubespray does something funky with an nginx reverse proxy/lb on each master node to avoid having an external LB for master nodes but i'll get to the bottom of this

**Comment by cyclinder** [comment]
> OK thank you all! do you have any idea where i can start troubleshooting this issue? what component is responsible to get the info from the api server and apply the corresponding configuration (ie, adding or removing ipvs rules) to the node? kubelet? kube-proxy directly?

kube-proxy will go to the add\delete\update ipvs rule according to the service's add\delete\update. 

for example, when creating a service, kube-proxy watch the addition of the service through api-server, and then kube-proxy adds the corresponding ipvs rules for this service in the node
